source/common/common/mem_block_builder.h

```diff
@@ -0,0 +1,126 @@
+// roughly as efficient as raw memcpy.
+template <class T> class MemBlockBuilder {
+public:
+  // Constructs a MemBlockBuilder of the specified size.
+  explicit MemBlockBuilder(uint64_t size)
+      : data_(std::make_unique<T[]>(size)), capacity_(size), write_ptr_(data_.get()) {}
+  MemBlockBuilder() : capacity_(0), write_ptr_(nullptr) {}
+
+  /**
+   * Populates (or repopulates) the MemBlockBuilder to make it the specified
+   * size. This does not have resize semantics; when populate() is called any
+   * previous contents are erased.
+   *
+   * @param size The number of memory elements to allocate.
+   */
+  void populate(uint64_t size) {
+    data_ = std::make_unique<T[]>(size);
+    capacity_ = size;
+    write_ptr_ = data_.get();
+  }
+
+  /**
+   * @return the capacity.
+   */
+  uint64_t capacity() const { return capacity_; }
+
+  /**
+   * Appends a single object of type T, moving an internal write-pointer
+   * forward. Asserts that there is room to write the object when compiled
+   * for debug.
+   *
+   * @param object the object to append.
+   */
+  void appendOne(T object) {
+    ASSERT(capacityRemaining() >= 1);
+    *write_ptr_++ = object;
+  }
+
+  /**
+   * Appends raw data, moving an internal write-pointer forward. Asserts
+   * that there is room to write the block when compiled for debug. It is
+   * the caller's responsibility to ensure that the input data is valid.
+   *
+   * @param data The block of objects to insert.
+   * @param size The number of elements in the block.
+   */
+  void appendData(const T* data, uint64_t size) {
+    ASSERT(capacityRemaining() >= size);
+    if (size != 0) {
+      memcpy(write_ptr_, data, size * sizeof(T));
+    }
+    write_ptr_ += size;
+  }
+
+  /**
+   * Appends the contents of another memory block to this one.
+   *
+   * @param src the block to append.
+   */
+  void appendBlock(const MemBlockBuilder& src) { appendData(src.data_.get(), src.size()); }
+
+  /**
+   * @return the number of elements remaining in the MemBlockBuilder.
+   */
+  uint64_t capacityRemaining() const { return (data_.get() + capacity_) - write_ptr_; }
+
+  /**
+   * Empties the contents of this.
+   */
+  void reset() {
+    data_.reset();
+    capacity_ = 0;
+    write_ptr_ = nullptr;
+  }
+
+  /**
+   * Returns the underlying storage as a unique pointer, clearing this.
+   *
+   * @return the transferred storage.
+   */
+  std::unique_ptr<T[]> release() {
+    write_ptr_ = nullptr;
+    capacity_ = 0;
+    return std::move(data_);
+  }
+
+  /**
+   * @return the populated data as a vector.
+   *
+   * This is exposed to help with unit testing.
+   */
+  std::vector<T> toVector() const { return std::vector<T>(data_.get(), write_ptr_); }
+
+  /**
+   * @return The number of elements the have been added to the builder.
+   */
+  uint64_t size() const { return write_ptr_ - data_.get(); }
+
+private:
+  std::unique_ptr<T[]> data_;
+  uint64_t capacity_;
+  T* write_ptr_;
```

Comment:
Would appreciate a comment here indicate that whenever it is not `nullptr` the `write_ptr_` is guaranteed to be within `data_`. This also seems like a reasonable thing to ASSERT. I continue to believe maybe the other assert should be a `RELEASE_ASSERT`.